query-node/mappings/src/label/tag.ts

```diff
@@ -0,0 +1,289 @@
+  }
+
+  const currentTagIds = (proposal.tags || []).map((t) => t.id)
+  const tagIdsToSet = _.union(currentTagIds, tagIds)
+  if (tagIdsToSet) {
+    const tags = await Promise.all(tagIdsToSet.map(async (tagId: string) => await getById(store, Tag, tagId)))
+    proposal.tags = tags.filter((t): t is Tag => !!t)
+    await store.save<Proposal>(proposal)
+  }
+
+  logger.info('new Tag is assigned to proposal', { tagIds, proposalId })
+  return proposal
+}
+
+export async function processUnassignTagsFromThread(
+  store: DatabaseManager,
+  metadata: DecodedMetadataObject<IUnassignTagsFromThread>,
+  isLead: boolean,
+  workerId: number
+): Promise<any> {
+  const { tagIds, threadId } = metadata
+
+  if (!isLead) {
+    const tagPermittedWorker: TagPermittedWorker | undefined = await getOneBy(store, TagPermittedWorker, {
+      workerId: workerId.toString(),
+    })
+    if (!tagPermittedWorker) {
+      return MetaprotocolTxError.TagPermNotAllowed
+    }
+  }
+
+  const forumThread: ForumThread | undefined = await getById(store, ForumThread, threadId)
+  if (!forumThread) {
+    return MetaprotocolTxError.TagInvalidThreadId
+  }
+
+  const remainedTags = (forumThread.newTags || []).filter((t) => !(tagIds || []).includes(t.id))
+  forumThread.newTags = remainedTags
+  await store.save<ForumThread>(forumThread)
+
+  logger.info('tag is unassigned from forumThread', { tagIds, threadId })
+  return forumThread
+}
+
+export async function processUnassignTagsFromProposal(
+  store: DatabaseManager,
+  metadata: DecodedMetadataObject<IUnassignTagsFromProposal>,
+  isLead: boolean,
+  workerId: number
+): Promise<any> {
+  const { tagIds, proposalId } = metadata
+
+  if (!isLead) {
+    const tagPermittedWorker: TagPermittedWorker | undefined = await getOneBy(store, TagPermittedWorker, {
+      workerId: workerId.toString(),
+    })
+    if (!tagPermittedWorker) {
+      return MetaprotocolTxError.TagPermNotAllowed
+    }
+  }
+
+  const proposal: Proposal | undefined = await getById(store, Proposal, proposalId)
+  if (!proposal) {
+    return MetaprotocolTxError.TagInvalidProposalId
+  }
+
+  const remainedTags = (proposal.tags || []).filter((t) => !(tagIds || []).includes(t.id))
+  proposal.tags = remainedTags
+  await store.save<Proposal>(proposal)
+
+  logger.info('tag is unassigned from proposal', { tagIds, proposalId })
+  return proposal
+}
+
+export async function processAllowTagToWorker(
+  store: DatabaseManager,
+  metadata: DecodedMetadataObject<IAllowTagToWorker>
+): Promise<any> {
+  const { workerId: assigneeId } = metadata
+
+  const tagPermittedWorker: TagPermittedWorker | undefined = await getOneBy(store, TagPermittedWorker, {
+    workerId: assigneeId.toString(),
+  })
+
+  if (!tagPermittedWorker) {
+    const newTagToWorker = new TagPermittedWorker({
+      workerId: assigneeId,
+    })
+    await store.save<TagPermittedWorker>(newTagToWorker)
+  }
+
+  logger.info('TagPermittedWorker has been allowed', { assigneeId })
```

Comment:
Let's move this log into the `if` above and add an `else` branch that will return an error indicating that the user already has permissions. Let's do a similar thing for disallow function